Port to Collabwrapper

Saumya-Mishra9129 opened the pull request:
Tested , collaboration works fine, both joiner and sharer share same grid, and state is also shared between them , but colors on both grid are different as they take colors from sugar. Do colors should be same?. 
@chimosky @quozl @JuiP @srevinsaju Please review.

Saumya-Mishra9129 committed:
Port to CollabWrapper

dbus.ExportedGObject class used by ChatTube has gone away
in Python 3, so in Python 3 we must port away from
ExportedGObject.

Telepathy Tubes is deprecated.  It is not essential to port away from
them, but it should be done eventually.

CollabWrapper does not depend on either.

Add CollabWrapper from https://github.com/sugarlabs/collabwrapper

srevinsaju commented on DeductoActivity.py:
```diff
@@ -44,21 +37,13 @@
-        except dbus.exceptions.DBusException as e:
-            _logger.error(str(e))
+
+        super(DeductoActivity, self).__init__(handle)
```
Am I right, if `super(DeductoActivity, self).__init__(handle)` can be simplified to `super().__init__(handle)` ?

Saumya-Mishra9129 replied:
Yeah you are , these arguments were necessary in Python 2 , but in Python 3  we can use either of them. Thanks

srevinsaju commented on DeductoActivity.py:
```diff
@@ -87,8 +72,13 @@ def __init__(self, handle):
         self._sharing = False
-        self._initiating = False
-        self._setup_presence_service()
+        self._initiating = None
```
May I know why this was changed?

Saumya-Mishra9129 replied:
Thanks for noticing , I find out that it was a mistake. I will change this one.

srevinsaju reviewed (commented):
did a code review. Nice! 

I am speculating about `set_data` and `get_data`. Are nothing to be done in those functions?

Saumya-Mishra9129 commented:
402f6b2 is not effective.

quozl commented:
Keep in mind that https://github.com/sugarlabs/colordeducto is very similar, and it saves effort to cherry-pick patches from one repository to the other.  The repositories have diverged slightly because people haven't kept that in mind.

quozl commented:
Thanks.

* [ ] collabwrapper.py differs from upstream,
* [ ] some of your changes don't pass style checks, e.g. space after comma,

Could not test, was unable to find a test case or procedure.  Can you suggest?

Saumya-Mishra9129 commented:
> Thanks.
> 
>  collabwrapper.py differs from upstream,
>  some of your changes don't pass style checks, e.g. space after comma,

Thanks @quozl for review. Fixed in 23863e9.

> Could not test, was unable to find a test case or procedure. Can you suggest?

There are the things I tried while testing collaboration -

- joiner is able to join the sharer.
- Clicking on Start a new game by sharer , starts a new game at joiner. Also only sharer is able to start a new game, joiner cannot.
- Same grid is shared between sharer and joiner ( same colours also).
- If joiner recognises a pattern and clicks on True or False buttons , game state will be updated in sharer and vice-versa.
-  Explore some examples is not shared between sharer and joiner.

quozl commented:
Thanks.  Thing is, I don't recognise patterns, and there's no instructions.  I haven't had time to read the source code.

Saumya-Mishra9129 commented:
> Thanks. Thing is, I don't recognise patterns, and there's no instructions. I haven't had time to read the source code.

@quozl I have looked at source code. There are 10 levels. Game will start from 1st level , if you give 5 correct answers, you will be updated to next level. These are the patterns for each level. 

- Level 1: Center dot is True
- Level 2: Corners match
- Level 3: more than 12 True
- Level 4: even number of True
- Level 5: diagonal is True
- Level 6: True above each False
- Level 7: True in each row
- Level 8: True path to center
- Level 9: more True on top half than on bottom half
- Level 10: smilely face

Here True is dark coloured dot and False is light coloured one.

quozl commented:
Thanks.  Tested.

* [ ] you've added a restriction that only the original sharer can press new game button; we don't usually implement this because it places a critical run-time dependency on the sharer's system and her engagement; if they leave and rejoin, nobody will be able to press new game button; we normally let any participant press new game button.
* [ ] the game state is not shared until new game button is pressed, see below.

@srevinsaju asked;
> I am speculating about `set_data` and `get_data`. Are nothing to be done in those functions?

These functions are to share the game state from a sharer to a joiner at the time the joiner joins.  Another method is for the sharer to send the grid when a joiner joins.

Saumya-Mishra9129 commented:
> you've added a restriction that only the original sharer can press new game button; we don't usually implement this because it places a critical run-time dependency on the sharer's system and her engagement; if they leave and rejoin, nobody will be able to press new game button; we normally let any participant press new game button.

Thanks , I haven't added this restriction. It was there already in master branch. But I can change it , remove this restriction.

> the game state is not shared until new game button is pressed, see below.

Thanks, I got it. Now I have to implement set_data and get_data.

Saumya-Mishra9129 committed:
Remove the restriction while collaborating : Only sharer can share a new game

Saumya-Mishra9129 committed:
Redraw sprites only when context is present

while collaboration a significant amount of delay can be seen during the updation of share state.

Saumya-Mishra9129 commented:
> the game state is not shared until new game button is pressed, see below.

Game state is shared without pressing new game button , but it had some delay. It takes time to update the shared state on receiver side.

quozl commented:
> Game state is shared without pressing new game button , but it had some delay. It takes time to update the shared state on receiver side.

Why is that?  Can you fix it?  I'm puzzled by the use of two joined signals from different levels of the stack; one for sugar3, and one for CollabWrapper.  It is improper to use CollabWrapper to transmit data before it has sent a joined signal, and the sugar3 joined signal may happen at a different time.

Saumya-Mishra9129 commented:
> Why is that?

I don't have any idea. The code is almost same as colourdeducto. However when I tested colordeducto in https://github.com/sugarlabs/colordeducto/pull/16 , It has less delay as compared to deducto. 

> Can you fix it? Can you fix it? I'm puzzled by the use of two joined signals from different levels of the stack; one for sugar3, and one for CollabWrapper. It is improper to use CollabWrapper to transmit data before it has sent a joined signal, and the sugar3 joined signal may happen at a different time.

I will try not to use two joined signal as you said. But I don't know it will fix this or not. This delay can be seen in many other activities as well.

srevinsaju commented:
You are right, colordeducto works faster than deducto, but looking at the diff, I am not able to make out why colordeducto works faster. On  colordeducto#16, you have also used two signals, like this one, but it works faster. Maybe colordeducto received another patch which deducto doesn't share?

quozl commented:
ImageViewer source demonstrates a minimal set of signals.  You might review that source.

To find the cause of the delay, you might add instrumentation (e.g. logging calls) and analyse the logs (sharer and joiner) against a tcpdump taken in parallel.

You might also use Coverage to make sure that you understand what lines are executed.

To resolve a theory that ColorDeducto code has a fix, compare the two source files across the repositories.  I use emacs ediff or meld.  There are some surprising differences in the respective _master_ branches in the `_new_tube_cb`.

Saumya-Mishra9129 commented:
> To resolve a theory that ColorDeducto code has a fix, compare the two source files across the repositories. I use emacs ediff or meld. There are some surprising differences in the respective master branches in the _new_tube_cb.

Thanks tried this but it couldn't help, actually delay is ambiguous. Sometimes I get same game state in just 10 seconds. Sometimes it takes more than 30 seconds.

quozl commented:
Do you see same delays using Chat activity?

Saumya-Mishra9129 commented:
> Do you see same delays using Chat activity?

Yeah I can see there. 10 seconds of delay and also Online , Offline , Connected timers. I also saw https://github.com/sugarlabs/sugar/issues/820 , You have mentioned here how ping times varies for Ubuntu.

quozl commented:
Use Linux kernel API tools (ps axww) to determine if any of the Telepathy processes terminate and restart during that test?  These processes are created after D-Bus receives IPC for Telepathy from Sugar, and should persist without PID change until Sugar logs out or restarts.

Saumya-Mishra9129 commented:
> Use Linux kernel API tools (ps axww) to determine if any of the Telepathy processes terminate and restart during that test? These processes are created after D-Bus receives IPC for Telepathy from Sugar, and should persist without PID change until Sugar logs out or restarts.

Thanks I will sure take a look.
Some interesting results can be seen while testing coverage.
**python3-coverage run /usr/bin/sugar-activity3**

```
1597044523.550005 ERROR deducto-activity: org.freedesktop.DBus.Error.NoReply: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
Traceback (most recent call last):
  File "/usr/bin/sugar-activity3", line 5, in <module>
    activityinstance.main()
  File "/usr/lib/python3.7/dist-packages/sugar3/activity/activityinstance.py", line 230, in main
    instance = create_activity_instance(activity_constructor, activity_handle)
  File "/usr/lib/python3.7/dist-packages/sugar3/activity/activityinstance.py", line 59, in create_activity_instance
    activity = constructor(handle)
  File "/usr/share/sugar/activities/deducto/DeductoActivity.py", line 62, in __init__
    self._setup_toolbars()
  File "/usr/share/sugar/activities/deducto/DeductoActivity.py", line 137, in _setup_toolbars
    stop_button = StopButton(self)
  File "/usr/lib/python3.7/dist-packages/sugar3/activity/widgets.py", line 92, in __init__
    activity.add_stop_button(self)
  File "/usr/lib/python3.7/dist-packages/sugar3/activity/activity.py", line 485, in add_stop_button
    self._stop_buttons.append(button)
AttributeError: 'DeductoActivity' object has no attribute '_stop_buttons'

```
It looks like a toolkit issue.

chimosky commented:
@Saumya-Mishra9129 the error you posted above is expected because on [L46](https://github.com/Saumya-Mishra9129/deducto/blob/major/DeductoActivity.py#L46) `super` is called and the `sugar3.activity.activity` class was never initialised.

Saumya-Mishra9129 commented:
> @Saumya-Mishra9129 the error you posted above is expected because on [L46](https://github.com/Saumya-Mishra9129/deducto/blob/major/DeductoActivity.py#L46) `super` is called and the `sugar3.activity.activity` class was never initialised.

I am sure whether this causes the issue or not. In Python 3  super(DeductoActivity, self).__init__(handle) can be simplified to super().__init__(handle) , both are same. I don't think this causes the error while testing coverage.

chimosky commented:
@Saumya-Mishra9129 said
> I am sure whether this causes the issue or not. In Python 3 super(DeductoActivity, self).init(handle) can be simplified to super().init(handle) , both are same. I don't think this causes the error while testing coverage.

I agree, it doesn't cause the error and I also don't think it's a toolkit issue.

chimosky commented:
Tested, previous error doesn't persist although there's a `PyGIWarning` in the logs,
```
PyGIWarning: PangoCairo was imported without specifying a version first. Use gi.require_version('PangoCairo', '1.0') before import to ensure that the right version gets loaded.
  from gi.repository import PangoCairo
```

Collaboration works as expected, although, the joiner's game colours change to the sharer's and is that intended?

quozl commented:
> AttributeError: 'DeductoActivity' object has no attribute '_stop_buttons'

This error is very clear evidence that the `__init__` has not run.  See https://github.com/sugarlabs/sugar-toolkit-gtk3/blob/68e57eff8458a89704bfc5b22f171514948bacd0/src/sugar3/activity/activity.py#L468

`_stop_buttons` is initialised by Activity's `__init__`, so for it not to have been run is very important to resolve.

_super_ doesn't have the same effect, in my experience, as calling the `__init__` using the class name.

In particular, we use multiple inheritance in Activity, and there are warnings I've found again just now which say _super_ is for single inheritance only.  Also, you cannot use _super_ if the subclasses don't use it, and vice versa.  Since our subclasses don't use _super_, that means we can't use it either.

* [Every occurrence of the method must include super() after you use it](https://www.pythonforbeginners.com/super/working-python-super-function).
* [Subclasses must use super if their superclasses do](https://fuhm.net/super-harmful/)

It would be impractical to change the toolkit to use _super_ because we would have to change all activities at once, and we wouldn't be able to use a new activity safely on an old Sugar, and vice versa.

Saumya-Mishra9129 commented:
> the joiner's game colours change to the sharer's
                                 
I think so , I have implemented collaboration that way here. @quozl and @chimosky Can you look I am not able to see any delay now. Collaboration works fine.

Saumya-Mishra9129 committed:
Remove super and use activity.Activity's __init__

we support multiple inheritance in activity, and super is for single inheritance only. We cannot use super in the subclasses if superclass doesn't use it

Saumya-Mishra9129 commented:
In particular, we use multiple inheritance in Activity, and there are warnings I've found again just now which say super is for single inheritance only. Also, you cannot use super if the subclasses don't use it, and vice versa. Since our subclasses don't use super, that means we can't use it either.

> Every occurrence of the method must include super() after you use it.
Subclasses must use super if their superclasses do
It would be impractical to change the toolkit to use super because we would have to change all activities at once, and we wouldn't be able to use a new activity safely on an old Sugar, and vice versa.

Thanks new info for me. I have removed super now.

quozl commented:
Thanks.  Reviewed again.  I've mentioned this before, but I'll put it another way;

* [ ] your code connects to both the Activity _joined_ signal and CollabWrapper's _joined_ signal, specifying a function that sets variables; the same function is connected to the Activity _shared_ signal; it isn't clear why your code does this, as CollabWrapper also connects to the Activity _joined_ signal and cascades to emitting the CollabWrapper _joined_ **or** CollabWrapper may emit the signal independently of the Activity _joined_ signal ... this is confusing; why do you want to connect these signals and is there another way you can do it?

Saumya-Mishra9129 commented:
> your code connects to both the Activity joined signal and CollabWrapper's joined signal, specifying a function that sets variables; the same function is connected to the Activity shared signal; it isn't clear why your code does this, as CollabWrapper also connects to the Activity joined signal and cascades to emitting the CollabWrapper joined or CollabWrapper may emit the signal independently of the Activity joined signal ... this is confusing; why do you want to connect these signals and is there another way you can do it?

Thanks , Can we not just use collabwrapper signals ? As CollabWrapper is used for collaboration, we should just call the joined event of collabwrapper and not the one from the sugar3.activity.activity.Activity class. 
Also I used this way because many activities I have seen implemented collaboration this way.

Saumya-Mishra9129 committed:
Remove call to activity joined signal

As CollabWrapper is used here for collaboration, use the joined event of collabwrapper only and not one from the sugar3.activity.Activity class.

Saumya-Mishra9129 committed:
Fix : Joiner can also initiate a new game

Saumya-Mishra9129 commented:
Tested till 485688b. Works Fine. Please review.
